url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/PartsRobotAgent.java

```diff
@@ -0,0 +1,531 @@
+
+	private enum PartsRobotStatus {
+		IDLE, PICKING_UP, PLACING
+	};
+
+	public Semaphore animation = new Semaphore(0, true);
+
+	// public Semaphore accessKit = new Semaphore(0, true);
+
+	public PartsRobotAgent(String name) {
+		super();
+
+		this.name = name;
+		time = new Date();
+		timer = new Timer();
+
+		// Add arms
+		for (int i = 0; i < 4; i++) {
+			this.Arms.add(new Arm());
+		}
+
+		status = PartsRobotStatus.IDLE;
+	}
+
+	/*
+	 * Messages
+	 */
+
+	@Override
+	public void msgSetDropChance(float dChance) {
+		print("Chance to drop is now set to: " + dChance);
+		dropChance = dChance;
+	}
+
+	/**
+	 * Changes the configuration for the kits From FCS
+	 */
+	@Override
+	public void msgHereIsKitConfiguration(KitConfig config) {
+		print("Received msgHereIsKitConfiguration");
+		Kitconfig = config;
+		kitsNum = 0;
+		GoodParts = new ConcurrentHashMap<Nest, List<Part>>();
+		// stateChanged();
+	}
+
+	/**
+	 * From Camera
+	 */
+	@Override
+	public void msgHereAreGoodParts(Nest n, List<Part> goodParts) {
+		print("Received msgHereAreGoodParts of type " + goodParts.get(0).type.getName());
+		GoodParts.put(n, goodParts);
+		print("I have " + MyKits.size() + " kits and " + GoodParts.size() + " nests");
+		stateChanged();
+	}
+
+	/**
+	 * From Stand
+	 */
+	@Override
+	public void msgUseThisKit(final Kit k) {
+		print("Received msgUseThisKit");
+
+		MyKit mk = new MyKit(k);
+		MyKits.add(mk);
+		print("I have " + MyKits.size() + " kits and " + GoodParts.size() + " nests");
+		stateChanged();
+
+	}
+
+	/**
+	 * Releases animation semaphore after a part is picked up, so that a new animation may be run by graphics. From
+	 * graphics
+	 */
+	@Override
+	public void msgPickUpPartDone() {
+		print("Received msgPickUpPartDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	/**
+	 * Releases animation semaphore after a part is given to kit, so that a new animation may be run by graphics. From
+	 * graphics
+	 */
+	@Override
+	public void msgGivePartToKitDone() {
+		print("Received msgGivePartToKitDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgDropPartFromArmDone() {
+		print("Received msgDropPartFromArmDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	/*
+	 * Scheduler
+	 * 
+	 * @see agent.Agent#pickAndExecuteAnAction()
+	 */
+
+	@Override
+	public boolean pickAndExecuteAnAction() {
+
+		// Checks if a kit is done and inspects it if it is
+		synchronized (MyKits) {
+			if (MyKits.size() > 0) {
+				for (MyKit mk : MyKits) {
+					if (mk.MKS == MyKitStatus.DONE) {
+						RequestInspection(mk);
+						return true;
+					}
+				}
+			}
+		}
+
+		// Checks if any arm is holding a part and places it if there is one
+		if (status == PartsRobotStatus.PLACING) {
+			synchronized (Arms) {
+				for (Arm arm : Arms) {
+					if (arm.AS == ArmStatus.FULL) {
+						print("Arm holding: " + arm.part.type.toString());
+						if (arm.part.type.getName() != "Dummy") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.